### PR TITLE
Add Horizontal and Vertical fields

### LINKED ISSUE #28: Declaration of Horizontal and Vertical Fields
It should be possible to declare also horizontal and vertical fields in dusk. @Stagno and me designed them as follows:

* Full: Field[Edge, K] / Field[Edge > Cell, K]
* Horizontal: Field[Edge] / Field[Edge > Cell]
* Vertical Field[K]

note the upper case K to clearly separate the concept from the commonly used iteration variable `k` for vertical regions. These declarations need to be possible both in arguments to the stencil as well as temporary fields. 

TODO:
- [x] update design document
- [x] implement

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

Adds possibility to declare horizontal and vertical fields. The following signatures were introduced:

* Full: Field[Edge, K] / Field[Edge > Cell, K]
* Horizontal: Field[Edge] / Field[Edge > Cell]
* Vertical Field[K]

## Resolves

Fixes #28

### CODE REVIEW

**Comment by BenWeber42** [comment]
Is this a blocker? What's the status of this syntax? Is it just a temporary solution or has it been discussed and _accepted_?

File: dusk/grammar.py [hunk 1]
```diff
@@ -144,24 +145,59 @@ def field_declaration(self, name: str, type: expr):
...
+            raise DuskSyntaxError(
+                "Field declaration needs to either specify horizontal or vertical dimensions"
+            )
+        if vindex is not None:
+            if vindex.id != "K":
+                raise DuskSyntaxError(
+                    f"Invalid field type, received vertical '{vindex.id}' but only K is allowed"
+                )
+        if hindex is None:
+            return None
```
Comment (BenWeber42) [inline dusk/grammar.py:197]:
```suggestion
        if hindex is None:
            return make_field_dimensions_vertical()
```
Why not `make_field_dimensions_vertical()`?
Then we could treat this _uniformly_ in `add_field_declaration`.

File: dusk/grammar.py [hunk 2]
```diff
@@ -144,24 +145,59 @@ def field_declaration(self, name: str, type: expr):
...
+                        ),
+                        ctx=Load,
+                    ),
+                    # FIXME: ensure built-ins (like `Edge`) aren't _shadowed_ by variables
+                    # TODO: hardcoded string
+                    Capture(OneOf(Compare, name(OneOf("Edge", "Cell", "Vertex")))).to(
+                        "hindex"
+                    ),
+                    Capture(Name).to("vindex"),
+                    None,
```
[the remark this answers is not in the document]
Comment (BenWeber42) [reply dusk/grammar.py:179]:
```suggestion
```

File: test.py [hunk 3]
```diff
@@ -222,3 +224,18 @@ def temp_field(a: Field[Edge], b: Field[Edge], out: Field[Edge]):
...
+    full: Field[Edge, K],
+    horizontal: Field[Edge],
+    horizontal_sparse: Field[Edge > Cell],
+    vertical: Field[K],
+):
+    with levels_downward as k:
+        out = full + horizontal + vertical
+        out = sum_over(
+            Edge > Cell, horizontal_sparse, weights=[]
+        )  # small bug in short reduce over causes weights to be requried, fix in seperate PR!
```
Comment (BenWeber42) [inline test.py:240]: Interesting. I'm gonna have to check this out.

File: dusk/grammar.py [hunk 4]
```diff
@@ -144,24 +145,59 @@ def field_declaration(self, name: str, type: expr):
...
+                            Capture(OneOf(Compare, Name)).to("hindex"),
+                            Capture(Name).to("vindex"),
+                        ),
+                        ctx=Load,
+                    ),
+                    # FIXME: ensure built-ins (like `Edge`) aren't _shadowed_ by variables
+                    # TODO: hardcoded string
+                    Capture(OneOf(Compare, name(OneOf("Edge", "Cell", "Vertex")))).to(
+                        "hindex"
+                    ),
```
Comment (BenWeber42) [inline dusk/grammar.py:177]: Just a note: we could consider some infrastructure to support _overloading_ or _optional positional arguments by type_ properly. This is the second time we encounter this (we also have this for field indexing). I feel like _overloading_ makes the code quite readable and intuitive, so we might use something like this more often in the future too.

File: test.py [hunk 5]
```diff
@@ -202,16 +204,20 @@ def other_vertical_iteration_variable(a: Field[Edge], b: Field[Edge]):
 
 
 @stencil
-def temp_field_demoted(a: Field[Edge], b: Field[Edge], out: Field[Edge]):
+def temp_field_demoted(a: Field[Edge, K], b: Field[Edge, K], out: Field[Edge, K]):
     x: Field[Edge]
+    y: Field[Edge, K]
     with levels_downward:
         x = a + b
```
Comment (BenWeber42) [inline test.py:211]: Interesting that dawn is happy with this. `x` doesn't have a vertical dimension. I guess it recognizes that this is actually a vertical region which has to be sequential.

**Comment by BenWeber42** [comment]
@mroethlin Please review & merge if you think it's fine.

File: test.py (hunk 5 above)
Comment (mroethlin) [reply test.py:211]: It's very debatable if this dusk code should be legal in my opinion. what happens in dawn is that x is demoted to a local variable
```
          for (auto const &loc : getEdges(LibTag{}, m_mesh)) {
            ::dawn::float_type __local_x_3_33 =
                (m_a(deref(LibTag{}, loc), k + 0) +
                 m_b(deref(LibTag{}, loc), k + 0));
            if ((__local_x_3_33 > (int)3)) {
              m_out(deref(LibTag{}, loc), k + 0) = __local_x_3_33;
            } else {
              m_out(deref(LibTag{}, loc), k + 0) = (int)5;
            }
          }
```
hence the codegen works out. An interesting case for sure.

Comment (BenWeber42) [reply test.py:211]: That's interesting. A temporary 2d field could be abused like so:
```python
def column_sum(x: Field[Edge, K]):
  sum: Field[Edge]
  with levels_upward[:0]: # bottom level
    sum = x
  with levels_upward[1:]: # remaining levels
    sum = sum + x
```
Would this work?

**Pull request merged**